Robot shooter

avidraccoon opened the pull request.

Commit:
Created basic state machine for shooter/intake subsystem

Commit:
Improved implementation of State Machine

Commit:
Fixed Shooter and Intake

Commit:
Simulation testing implementation

jkleiber commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -19,6 +21,17 @@ public class RobotContainer {
     private Supplier<Pose2d> simulatedPoseSupplier;
+    private ShooterIntakeSubsystem intakeSubsystem =
+            new ShooterIntakeSubsystem(
+                    new ShooterIntakeIO() {
+
+                        @Override
+                        public void setVoltage(double volts) {
+                            // TODO Auto-generated method stub
+                            throw new UnsupportedOperationException(
+                                    "Unimplemented method 'setVoltage'");
+                        }
+                    });
```
We should be implementing the IO classes in separate files - one for hardware and one for sim. They both should use the interface you've created.

avidraccoon replied:
Forgot to push new commit

avidraccoon replied:
I just forgot to remove those lines of code I believe.

jkleiber commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -68,7 +79,14 @@ public void robotPeriodic() {
         // block in order for anything in the Command-based framework to work.
+
         CommandScheduler.getInstance().run();
+        Logger.recordOutput("randomCount", pose2d);
+        Transform2d modifier = transform2d.times(Math.sin(count / 100.0));
+        pose2d = pose2d.transformBy(modifier);
+        Logger.recordOutput("joystickExpected", Math.sin(count / 100.0));
+
+        count++;
```
Is this just test code?

avidraccoon replied:
Yes let me remove it quickly.

jkleiber commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -24,6 +28,11 @@ public class Robot extends LoggedRobot {
 
     private RobotContainer robotContainer;
 
+    Pose2d pose2d = new Pose2d(0.0, 0.0, new Rotation2d(0.0, 0.0));
+    Transform2d transform2d =
+            new Transform2d(new Translation2d(0.02, 0.0), new Rotation2d(0.0, 0.0));
+    int count = 0;
+
```
Do you need pose for the shooter?

avidraccoon replied:
No sorry

jkleiber commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -52,6 +61,8 @@ public void robotInit() {
         }
 
         Logger.start();
+        Logger.recordOutput("randomTesting", new Pose2d(0.0, 0.0, new Rotation2d(0.0, 0.0)));
```
can we remove this test code before we merge this?

avidraccoon replied:
Yes

jkleiber commented on src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    // spotless:off
+    public enum State {
```
After looking at this more, I'm not sure using the enum functions really is buying us much. It's nice that we get some inheritance here, but it's also sort of confusing to read with `changeStates` and `changeState` having two different outcomes.

One thing that I think is unfortunate is that we have to have a static instance of the subsystem inside the enum. This is not intuitive and I don't think we want to design things this way.

I think we should go back to the system we did last year for now - an enum that defines state and a switch statement that dispatches specific functions which each run the logic for their respective state.

avidraccoon replied:
I believe I can fix both of those issue very quickly if there is not other issues is it fine to keep this implementation with a bit of cleaning up.

jkleiber reviewed (commented):
First pass review of the shooter

Commit:
Simulation IO and cleanup

avidraccoon commented:
I'm having problems with solving the conflicting files can you help me?

jkleiber commented:
@avidraccoon what I do when I see merge conflicts is try to preserve the code that is in conflict with my code while still making sure my feature works. In this case it probably makes sense to "Accept Both" changes when doing the rebase and then refactor the code as needed to make things work.

The rebase process can be found here: https://team401.github.io/Docs/Programming/Software-Workflow.html#keep-branch-updated

If there's a specific problem I can help with just let me know!

jkleiber commented on src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java:
```diff
@@ -0,0 +1,145 @@
+
+public class ShooterIntakeSubsystem extends SubsystemBase {
+
+    // Move to constants later
+    double intakeTargetRPM = -100.0;
+    double shootingTargetRPM = 300.0;
+    //
```
Let's move these to constants

We can store constants in src/main/java/frc/robot/constants and eventually JSONify everything

jkleiber commented on src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeIOSim.java:
```diff
@@ -0,0 +1,57 @@
+public class ShooterIntakeIOSim implements ShooterIntakeIO {
+
+    FlywheelSim flywheelSim = new FlywheelSim(DCMotor.getCIM(2), 1.0, 1.0);
+    PIDController controller = new PIDController(0.75, 0.1, 0.0);
```
If these are going to be the control gains, we should move them to constants as well

jkleiber reviewed (commented):
Second pass, this looks pretty clean now. We just need to get the merge conflicts resolved and then it'll be ready to merge.

After we merge this, we need to get the hardware IO made so we can test on the robot. Can you create a ticket to do this if you haven't already?

Commit:
Fixed naming and logging

Commit:
Merge branch 'main' into Robot-Shooter

Commit:
add missing imports

Commit:
Moved Constants (Does not build)

avidraccoon commented:
I moved the constants and prepared for merge but now it can't build because it can find classes for import, how do I fix this?

Edit: It can build but simulation will not run.

Edit 2: Fixed it just needed to clear the build cache

jkleiber reviewed (approved):
There's some refactoring to do here, but let's do that in a separate issue. Approved - please merge when you are able
